File: examples/14-identity-and-privilege/cpp/src/main.cpp
// pmon v3 — identity & privilege (chapter 14).
//
// Subcommands:
//   pmon drop --user <name> [--keep-cap net_bind_service] -- CMD [args...]
//   pmon bindprobe [--port 80]
//
// `drop` runs as root and hands CMD to an UNPRIVILEGED user, optionally
// carrying exactly one capability (CAP_NET_BIND_SERVICE) across the identity
// change and the following execve(2) via the ambient-capability mechanism.
// The whole lesson is the ORDER of operations:
//
//   1. prctl(PR_SET_KEEPCAPS, 1)   — before the uid change, so the permitted
//                                    set survives setresuid(2) instead of
//                                    being scrubbed when uid goes 0 -> !0.
//   2. setgroups / setresgid / setresuid   — become the target user.
//   3. capset(2)                   — pin CAP_NET_BIND_SERVICE into the
//                                    permitted AND inheritable sets (ambient
//                                    requires both), dropping everything else.
//   4. prctl(PR_CAP_AMBIENT_RAISE) — raise it into the ambient set, the only
//                                    set that a non-root, non-file-cap execve
//                                    preserves.
//   5. execvp(2)                   — CMD now runs as <name> yet keeps :80.
//
// Raw prctl(2) / capset(2) / setres*id(2) on purpose — no libcap. RAII owns
// the socket; std::expected carries syscall failures; nothing is handled the
// C way (no scattered perror/exit). Deliberately single-threaded: a privilege
// drop must touch every task, and execve() only makes sense from one thread.

#include <array>
#include <cerrno>
#include <cstdint>
#include <cstdlib>
#include <cstring>
#include <expected>
#include <print>
#include <span>
#include <string>
#include <string_view>
#include <vector>

#include <arpa/inet.h>
#include <grp.h>
#include <linux/capability.h>
#include <netinet/in.h>
#include <pwd.h>
#include <sys/prctl.h>
#include <sys/socket.h>
#include <sys/syscall.h>
#include <unistd.h>

namespace {

// A syscall failure: what we were attempting, plus the captured errno.
struct SysErr {
  std::string what;
  int err;
};

// Wrap a raw syscall return: <0 becomes an error carrying errno.
std::expected<void, SysErr> checked(long rc, std::string_view what) {
  if (rc < 0) {
    return std::unexpected(SysErr{std::string(what), errno});
  }
  return {};
}

// RAII owner for a file descriptor — closed exactly once on scope exit.
class Fd {
 public:
  explicit Fd(int fd) noexcept : fd_(fd) {}
  ~Fd() {
    if (fd_ >= 0) ::close(fd_);
  }
  Fd(const Fd&) = delete;
  Fd& operator=(const Fd&) = delete;
  Fd(Fd&& other) noexcept : fd_(other.fd_) { other.fd_ = -1; }
  Fd& operator=(Fd&&) = delete;

  [[nodiscard]] int get() const noexcept { return fd_; }
  [[nodiscard]] bool valid() const noexcept { return fd_ >= 0; }

 private:
  int fd_ = -1;
};

// ---- bindprobe -----------------------------------------------------------
//
// Bind a TCP socket on <port> and report the effective uid. The whole point:
// binding a port < 1024 is impossible for a non-root process UNLESS it holds
// CAP_NET_BIND_SERVICE. Success at uid!=0 on :80 is the proof the ambient cap
// survived. Failure prints a stable "Permission denied" line and exits 3.
int cmd_bindprobe(std::span<char*> args) {
  int port = 80;
  for (std::size_t i = 0; i < args.size(); ++i) {
    std::string_view a = args[i];
    if (a == "--port" && i + 1 < args.size()) {
      port = std::atoi(args[++i]);
    } else {
      std::println(stderr, "bindprobe: unexpected argument: {}", a);
      return 2;
    }
  }

  Fd sock(::socket(AF_INET, SOCK_STREAM, 0));
  if (!sock.valid()) {
    std::println(stderr, "bindprobe: socket: {}", std::strerror(errno));
    return 1;
  }
  int one = 1;
  ::setsockopt(sock.get(), SOL_SOCKET, SO_REUSEADDR, &one, sizeof(one));

  sockaddr_in addr{};
  addr.sin_family = AF_INET;
  addr.sin_addr.s_addr = htonl(INADDR_ANY);
  addr.sin_port = htons(static_cast<uint16_t>(port));

  if (::bind(sock.get(), reinterpret_cast<sockaddr*>(&addr), sizeof(addr)) != 0) {
    std::println(stderr, "bindprobe: bind :{}: {}", port, std::strerror(errno));
    return 3;
  }
  std::println("bindprobe: uid={} bound :{}", ::getuid(), port);
  return 0;
}

// ---- drop ----------------------------------------------------------------

// Perform the identity change (and optional cap retention), then execvp CMD.
// Returns only on failure; on success control never comes back.
std::expected<void, SysErr> arrange_and_exec(const passwd& pw, bool keep_cap,
                                             std::span<char*> cmd) {
  // (1) Retain the permitted set across the coming uid change. Must precede
  //     setresuid — after it, it is too late.
  if (keep_cap) {
    if (auto r = checked(::prctl(PR_SET_KEEPCAPS, 1, 0, 0, 0), "PR_SET_KEEPCAPS");
        !r)
      return r;
  }

  // (2) Become the target user: groups first, gid before uid — once uid drops
  //     we can no longer change gids.
  gid_t gid = pw.pw_gid;
  if (auto r = checked(::setgroups(1, &gid), "setgroups"); !r) return r;
  if (auto r = checked(::setresgid(gid, gid, gid), "setresgid"); !r) return r;
  if (auto r = checked(::setresuid(pw.pw_uid, pw.pw_uid, pw.pw_uid), "setresuid");
      !r)
    return r;

  // (3) Pin exactly CAP_NET_BIND_SERVICE into permitted+inheritable (ambient
  //     demands the cap live in both), dropping every other retained cap.
  if (keep_cap) {
    __user_cap_header_struct hdr{};
    hdr.version = _LINUX_CAPABILITY_VERSION_3;
    hdr.pid = 0;
    std::array<__user_cap_data_struct, 2> data{};
    const std::uint32_t bit = 1u << CAP_NET_BIND_SERVICE;  // cap 10, word 0
    data[0].effective = bit;
    data[0].permitted = bit;
    data[0].inheritable = bit;
    if (auto r = checked(::syscall(SYS_capset, &hdr, data.data()), "capset"); !r)
      return r;

    // (4) Raise it into the ambient set — the only set an unprivileged execve
    //     of a file without file-caps carries forward.
    if (auto r = checked(::prctl(PR_CAP_AMBIENT, PR_CAP_AMBIENT_RAISE,
                                 CAP_NET_BIND_SERVICE, 0, 0),
                         "PR_CAP_AMBIENT_RAISE");
        !r)
      return r;
  }

  // (5) Hand off. CMD now runs as <name>, keeping :80 only if we raised it.
  std::vector<char*> args(cmd.begin(), cmd.end());
  args.push_back(nullptr);
  ::execvp(args[0], args.data());
  return std::unexpected(SysErr{std::string("execvp ") + args[0], errno});
}

int cmd_drop(std::span<char*> args) {
  std::string_view user;
  bool keep_cap = false;
  std::span<char*> cmd;
  for (std::size_t i = 0; i < args.size(); ++i) {
    std::string_view a = args[i];
    if (a == "--user" && i + 1 < args.size()) {
      user = args[++i];
    } else if (a == "--keep-cap" && i + 1 < args.size()) {
      std::string_view cap = args[++i];
      if (cap != "net_bind_service") {
        std::println(stderr, "drop: unsupported --keep-cap: {}", cap);
        return 2;
      }
      keep_cap = true;
    } else if (a == "--") {
      cmd = args.subspan(i + 1);
      break;
    } else {
      std::println(stderr, "drop: unexpected argument: {}", a);
      return 2;
    }
  }
  if (user.empty()) {
    std::println(stderr, "drop: --user <name> is required");
    return 2;
  }
  if (cmd.empty()) {
    std::println(stderr, "drop: missing -- CMD");
    return 2;
  }
  if (::getuid() != 0) {
    std::println(stderr, "drop: must run as root");
    return 1;
  }

  errno = 0;
  passwd* pw = ::getpwnam(std::string(user).c_str());
  if (pw == nullptr) {
    std::println(stderr, "drop: unknown user: {}", user);
    return 1;
  }

  auto r = arrange_and_exec(*pw, keep_cap, cmd);
  // Only reached if a step failed (exec never returns on success).
  std::println(stderr, "drop: {}: {}", r.error().what, std::strerror(r.error().err));
  return 1;
}

void usage() {
  std::println(stderr,
               "usage:\n"
               "  pmon drop --user <name> [--keep-cap net_bind_service] -- CMD [args...]\n"
               "  pmon bindprobe [--port 80]");
}

}  // namespace

int main(int argc, char** argv) {
  std::span<char*> all(argv, static_cast<std::size_t>(argc));
  if (all.size() < 2) {
    usage();
    return 2;
  }
  std::string_view sub = all[1];
  std::span<char*> rest = all.subspan(2);
  if (sub == "drop") return cmd_drop(rest);
  if (sub == "bindprobe") return cmd_bindprobe(rest);
  usage();
  return 2;
}

File: examples/14-identity-and-privilege/rust/src/main.rs
//! pmon v3 — identity & privilege (chapter 14).
//!
//! Subcommands:
//!   pmon drop --user <name> [--keep-cap net_bind_service] -- CMD [args...]
//!   pmon bindprobe [--port 80]
//!
//! `drop` runs as root and hands CMD to an unprivileged user, optionally
//! carrying exactly one capability (CAP_NET_BIND_SERVICE) across the identity
//! change and the following execvp(3) via the ambient-capability mechanism.
//! The ordering is the lesson (same as the C++ raw path):
//!
//! 1. prctl(PR_SET_KEEPCAPS, 1) before the uid change, so the permitted set
//!    survives setresuid(2).
//! 2. setgroups / setresgid / setresuid become the target user.
//! 3. capset(2) pins CAP_NET_BIND_SERVICE into permitted + inheritable
//!    (ambient requires both).
//! 4. prctl(PR_CAP_AMBIENT_RAISE) raises it into the ambient set — the only
//!    set a non-root, no-file-cap execve keeps.
//! 5. execvp(3): CMD runs as <name>, keeping :80 iff the cap was raised.
//!
//! Raw libc sequence, mirroring the C++ build; std OwnedFd gives the socket
//! RAII; errors flow through Result/?.

use std::ffi::{CStr, CString, c_char, c_void};
use std::os::fd::{AsRawFd, FromRawFd, OwnedFd};

// Not exported by the libc crate; both are stable kernel ABI constants.
const LINUX_CAPABILITY_VERSION_3: u32 = 0x2008_0522;
const CAP_NET_BIND_SERVICE: u32 = 10;

/// A syscall failure: what we attempted, plus the captured errno.
struct SysErr {
    what: String,
    errno: i32,
}

/// libc strerror(3), so the text matches the C++/Go builds byte-for-byte.
fn strerror(errno: i32) -> String {
    // SAFETY: strerror returns a valid NUL-terminated static string.
    unsafe { CStr::from_ptr(libc::strerror(errno)) }
        .to_string_lossy()
        .into_owned()
}

fn last_errno() -> i32 {
    std::io::Error::last_os_error().raw_os_error().unwrap_or(0)
}

/// Turn a negative syscall return into an error carrying errno.
fn check(rc: i64, what: &str) -> Result<(), SysErr> {
    if rc < 0 {
        Err(SysErr {
            what: what.to_string(),
            errno: last_errno(),
        })
    } else {
        Ok(())
    }
}

// ---- bindprobe -----------------------------------------------------------

fn cmd_bindprobe(args: &[String]) -> i32 {
    let mut port: u16 = 80;
    let mut i = 0;
    while i < args.len() {
        match args[i].as_str() {
            "--port" => {
                let Some(v) = args.get(i + 1) else {
                    eprintln!("bindprobe: --port needs a value");
                    return 2;
                };
                match v.parse::<u16>() {
                    Ok(p) => port = p,
                    Err(_) => {
                        eprintln!("bindprobe: bad port: {v}");
                        return 2;
                    }
                }
                i += 1;
            }
            other => {
                eprintln!("bindprobe: unexpected argument: {other}");
                return 2;
            }
        }
        i += 1;
    }

    // SAFETY: raw socket then immediately owned; all pointers below are to
    // locals whose lifetime covers each call.
    let raw = unsafe { libc::socket(libc::AF_INET, libc::SOCK_STREAM, 0) };
    if raw < 0 {
        eprintln!("bindprobe: socket: {}", strerror(last_errno()));
        return 1;
    }
    let sock = unsafe { OwnedFd::from_raw_fd(raw) };

    let one: i32 = 1;
    unsafe {
        libc::setsockopt(
            sock.as_raw_fd(),
            libc::SOL_SOCKET,
            libc::SO_REUSEADDR,
            &one as *const i32 as *const c_void,
            std::mem::size_of::<i32>() as libc::socklen_t,
        );
    }

    let mut addr: libc::sockaddr_in = unsafe { std::mem::zeroed() };
    addr.sin_family = libc::AF_INET as libc::sa_family_t;
    addr.sin_addr.s_addr = libc::INADDR_ANY.to_be();
    addr.sin_port = port.to_be();

    let rc = unsafe {
        libc::bind(
            sock.as_raw_fd(),
            &addr as *const libc::sockaddr_in as *const libc::sockaddr,
            std::mem::size_of::<libc::sockaddr_in>() as libc::socklen_t,
        )
    };
    if rc != 0 {
        eprintln!("bindprobe: bind :{port}: {}", strerror(last_errno()));
        return 3;
    }
    let uid = unsafe { libc::getuid() };
    println!("bindprobe: uid={uid} bound :{port}");
    0
}

// ---- drop ----------------------------------------------------------------

#[repr(C)]
struct CapHeader {
    version: u32,
    pid: i32,
}

#[repr(C)]
#[derive(Clone, Copy)]
struct CapData {
    effective: u32,
    permitted: u32,
    inheritable: u32,
}

/// Drop to `uid`/`gid`, optionally retaining CAP_NET_BIND_SERVICE, then exec.
/// Returns only on failure; on success control never comes back.
fn arrange_and_exec(
    uid: libc::uid_t,
    gid: libc::gid_t,
    keep_cap: bool,
    cmd: &[String],
) -> Result<(), SysErr> {
    // (1) Retain the permitted set across the uid change (must precede it).
    if keep_cap {
        check(
            unsafe { libc::prctl(libc::PR_SET_KEEPCAPS, 1, 0, 0, 0) } as i64,
            "PR_SET_KEEPCAPS",
        )?;
    }

    // (2) Become the target user: groups, then gid, then uid.
    check(unsafe { libc::setgroups(1, &gid) } as i64, "setgroups")?;
    check(
        unsafe { libc::setresgid(gid, gid, gid) } as i64,
        "setresgid",
    )?;
    check(
        unsafe { libc::setresuid(uid, uid, uid) } as i64,
        "setresuid",
    )?;

    // (3) Pin exactly CAP_NET_BIND_SERVICE into permitted+inheritable.
    if keep_cap {
        let hdr = CapHeader {
            version: LINUX_CAPABILITY_VERSION_3,
            pid: 0,
        };
        let bit = 1u32 << CAP_NET_BIND_SERVICE; // cap 10, word 0
        let data = [
            CapData {
                effective: bit,
                permitted: bit,
                inheritable: bit,
            },
            CapData {
                effective: 0,
                permitted: 0,
                inheritable: 0,
            },
        ];
        check(
            unsafe { libc::syscall(libc::SYS_capset, &hdr as *const CapHeader, data.as_ptr()) },
            "capset",
        )?;

        // (4) Raise it into the ambient set.
        check(
            unsafe {
                libc::prctl(
                    libc::PR_CAP_AMBIENT,
                    libc::PR_CAP_AMBIENT_RAISE as libc::c_ulong,
                    CAP_NET_BIND_SERVICE as libc::c_ulong,
                    0,
                    0,
                )
            } as i64,
            "PR_CAP_AMBIENT_RAISE",
        )?;
    }

    // (5) Hand off to CMD.
    let prog = CString::new(cmd[0].as_str()).map_err(|_| SysErr {
        what: "bad CMD".into(),
        errno: libc::EINVAL,
    })?;
    let cargs: Vec<CString> = cmd
        .iter()
        .map(|s| CString::new(s.as_str()).unwrap_or_default())
        .collect();
    let mut ptrs: Vec<*const c_char> = cargs.iter().map(|c| c.as_ptr()).collect();
    ptrs.push(std::ptr::null());
    unsafe { libc::execvp(prog.as_ptr(), ptrs.as_ptr()) };
    Err(SysErr {
        what: format!("execvp {}", cmd[0]),
        errno: last_errno(),
    })
}

fn cmd_drop(args: &[String]) -> i32 {
    let mut user: Option<&str> = None;
    let mut keep_cap = false;
    let mut cmd: Vec<String> = Vec::new();
    let mut i = 0;
    while i < args.len() {
        match args[i].as_str() {
            "--user" => {
                let Some(v) = args.get(i + 1) else {
                    eprintln!("drop: --user needs a value");
                    return 2;
                };
                user = Some(v);
                i += 1;
            }
            "--keep-cap" => {
                let Some(v) = args.get(i + 1) else {
                    eprintln!("drop: --keep-cap needs a value");
                    return 2;
                };
                if v != "net_bind_service" {
                    eprintln!("drop: unsupported --keep-cap: {v}");
                    return 2;
                }
                keep_cap = true;
                i += 1;
            }
            "--" => {
                cmd = args[i + 1..].to_vec();
                break;
            }
            other => {
                eprintln!("drop: unexpected argument: {other}");
                return 2;
            }
        }
        i += 1;
    }

    let Some(user) = user else {
        eprintln!("drop: --user <name> is required");
        return 2;
    };
    if cmd.is_empty() {
        eprintln!("drop: missing -- CMD");
        return 2;
    }
    if unsafe { libc::getuid() } != 0 {
        eprintln!("drop: must run as root");
        return 1;
    }

    let cname = match CString::new(user) {
        Ok(c) => c,
        Err(_) => {
            eprintln!("drop: unknown user: {user}");
            return 1;
        }
    };
    // SAFETY: getpwnam returns a pointer into a static buffer valid until the
    // next call; we read it immediately.
    let pw = unsafe { libc::getpwnam(cname.as_ptr()) };
    if pw.is_null() {
        eprintln!("drop: unknown user: {user}");
        return 1;
    }
    let (uid, gid) = unsafe { ((*pw).pw_uid, (*pw).pw_gid) };

    match arrange_and_exec(uid, gid, keep_cap, &cmd) {
        Ok(()) => 0, // unreachable: exec never returns on success
        Err(e) => {
            eprintln!("drop: {}: {}", e.what, strerror(e.errno));
            1
        }
    }
}

fn usage() {
    eprintln!("usage:");
    eprintln!("  pmon drop --user <name> [--keep-cap net_bind_service] -- CMD [args...]");
    eprintln!("  pmon bindprobe [--port 80]");
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        usage();
        std::process::exit(2);
    }
    let code = match args[1].as_str() {
        "drop" => cmd_drop(&args[2..]),
        "bindprobe" => cmd_bindprobe(&args[2..]),
        _ => {
            usage();
            2
        }
    };
    std::process::exit(code);
}
